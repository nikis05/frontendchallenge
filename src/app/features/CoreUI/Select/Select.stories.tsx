import React, { useState } from 'react';
import { action } from '@storybook/addon-actions';

import Select from './Select';
import '../../App/Root/Root.scss';

export default {
  component: Select,
  title: 'Select',
};

export const normal = () => (
  <Select
    options={[
      { key: '1', label: 'One' },
      { key: '2', label: 'Two' },
      { key: '3', label: 'Three' },
    ]}
    value='1'
    onChange={action('change')}
  />
);

export const multiple = () => (
  <Select
    options={[
      { key: '1', label: 'One' },
      { key: '2', label: 'Two' },
      { key: '3', label: 'Three' },
    ]}
    value={['1', '2']}
    onChange={action('change')}
  />
);

const Controller: React.FC<{ multipleMode: boolean }> = ({ multipleMode }) => {
  const [state, setState] = useState<string | string[] | null>(
    multipleMode ? [] : null,
  );
  return (
    <Select
      options={[
        { key: '1', label: 'One' },
        { key: '2', label: 'Two' },
        { key: '3', label: 'Three' },
      ]}
      value={state}
      onChange={setState}
    />
  );
};

export const controlledSingle = () => <Controller multipleMode={false} />;

export const controlledMultiple = () => <Controller multipleMode />;
