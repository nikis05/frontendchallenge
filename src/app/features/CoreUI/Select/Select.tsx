import React, { useState, FC } from 'react';

import DropDownMenu from '../DropDownMenu/DropDownMenu';
import { MenuOption } from '../DropDownMenu/Option/Option';

interface Props {
  /**
   * A list of Select options
   */
  options: MenuOption[];

  /**
   * Placeholder to show when no ouption is selected
   */
  placeholder?: string;

  /**
   * Current value of Select
   */
  value: string | null | string[];

  /**
   * Register callback for change event
   */
  onChange: (value: string | null | string[]) => void;
}

// Returns current value of select as display string. For empty value falls back
// to placeholder if provided. Otherwise returns default placeholder.
const getMenuLabel = (
  options: MenuOption[],
  value: string | null | string[],
  customPlaceholder: string | undefined,
) => {
  const getOptionDisplayName = (key: string) => options.find((option) => option.key === key)!.label;
  const placeholder = customPlaceholder || 'Select an option';
  if (Array.isArray(value)) {
    if (value.length === 0) return placeholder;
    return value.map(getOptionDisplayName).join(', ');
  }
  if (value === null) return placeholder;
  return getOptionDisplayName(value);
};

const removeFromArray = (array: string[], key: string) => array.filter((existingKey) => existingKey !== key);

const addToArray = (array: string[], key: string) => [...array, key];

const Select: FC<Props> = ({ options, placeholder, value, onChange }) => {
  const [isOpen, setOpen] = useState(false);
  const isMultipleMode = Array.isArray(value);

  const handleSelect = (key: string) => {
    // In multiple mode, calls onChange with updated array of selected keys
    if (isMultipleMode) {
      // Typescript type assertion
      const valueAsArray = value as string[];
      // If key is currently present in array, remove it from array
      if (valueAsArray.includes(key)) onChange(removeFromArray(valueAsArray, key));
      // Otherwise add it to array
      else onChange(addToArray(valueAsArray, key));
    } else onChange(key);
  };

  const menuLabel = getMenuLabel(options, value, placeholder);
  const selectedKeys = Array.isArray(value) ? value : undefined;

  return (
    <DropDownMenu
      label={menuLabel}
      options={options}
      isOpen={isOpen}
      onOpen={setOpen}
      selectedKeys={selectedKeys}
      onSelect={handleSelect}
      noCloseOnSelect={isMultipleMode}
    />
  );
};

export default Select;
