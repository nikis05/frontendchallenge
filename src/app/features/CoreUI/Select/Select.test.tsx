import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';

import Select from './Select';

describe('Select', () => {
  afterEach(cleanup);

  it('supports single select', () => {
    const onChange = jest.fn();
    const { container } = render(
      <Select
        placeholder='placeholder'
        options={[
          { key: '1', label: 'One' },
          { key: '2', label: 'Two' },
          { key: '3', label: 'Three' },
        ]}
        value='1'
        onChange={onChange}
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
    const selectionElement = container.getElementsByClassName('selection')[0];
    fireEvent.click(selectionElement);
    const optionElement = container.getElementsByClassName('option')[2];
    fireEvent.click(optionElement);
    expect(onChange).toBeCalledWith('3');
  });

  it('supports multi select', () => {
    const onChange = jest.fn();
    const { container } = render(
      <Select
        placeholder='placeholder'
        options={[
          { key: '1', label: 'One' },
          { key: '2', label: 'Two' },
          { key: '3', label: 'Three' },
        ]}
        value={['1', '3']}
        onChange={onChange}
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
    const selectionElement = container.getElementsByClassName('selection')[0];
    fireEvent.click(selectionElement);
    const optionElement = container.getElementsByClassName('option')[1];
    fireEvent.click(optionElement);
    expect(onChange).toBeCalledWith(['1', '3', '2']);
  });

  it('supports option deselect', () => {
    const onChange = jest.fn();
    const { container } = render(
      <Select
        placeholder='placeholder'
        options={[
          { key: '1', label: 'One' },
          { key: '2', label: 'Two' },
          { key: '3', label: 'Three' },
        ]}
        value={['1', '3']}
        onChange={onChange}
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
    const selectionElement = container.getElementsByClassName('selection')[0];
    fireEvent.click(selectionElement);
    const optionElement = container.getElementsByClassName('option')[2];
    fireEvent.click(optionElement);
    expect(onChange).toBeCalledWith(['1']);
  });
});
