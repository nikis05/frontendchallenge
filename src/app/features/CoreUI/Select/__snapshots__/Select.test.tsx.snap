// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select supports multi select 1`] = `
<div
  class="container"
>
  <div
    class="selection"
    tabindex="0"
  >
    <label
      class="container inline"
      role="presentation"
    >
      <div
        class="label"
      >
        One, Three
      </div>
    </label>
    <svg
      class="arrow"
      fill="none"
      height="4"
      viewBox="0 0 8 4"
      width="8"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M4 4L0 0L8 0L4 4Z"
        fill="#a5b1c4"
      />
    </svg>
  </div>
  <div
    class="wrapper"
  >
    <div
      class="option"
      tabindex="0"
    >
      <div
        class="container"
      >
        <label
          class="container right"
          role="presentation"
        >
          <div
            class="label"
          >
            One
          </div>
          <svg
            fill="none"
            height="20"
            viewBox="0 0 20 20"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3.33333 2.5H16.6667C16.8877 2.5 17.0996 2.5878 17.2559 2.74408C17.4122 2.90036 17.5 3.11232 17.5 3.33333V16.6667C17.5 16.8877 17.4122 17.0996 17.2559 17.2559C17.0996 17.4122 16.8877 17.5 16.6667 17.5H3.33333C3.11232 17.5 2.90036 17.4122 2.74408 17.2559C2.5878 17.0996 2.5 16.8877 2.5 16.6667V3.33333C2.5 3.11232 2.5878 2.90036 2.74408 2.74408C2.90036 2.5878 3.11232 2.5 3.33333 2.5ZM9.16917 13.3333L15.0608 7.44083L13.8825 6.2625L9.16917 10.9767L6.81167 8.61917L5.63333 9.7975L9.16917 13.3333Z"
              fill="#3385FF"
            />
          </svg>
          <input
            checked=""
            class="input"
            type="checkbox"
          />
        </label>
      </div>
    </div>
    <div
      class="option"
      tabindex="0"
    >
      <div
        class="container"
      >
        <label
          class="container right"
          role="presentation"
        >
          <div
            class="label"
          >
            Two
          </div>
          <svg
            fill="none"
            height="20"
            viewBox="0 0 20 20"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3.33333 2.5H16.6667C16.8877 2.5 17.0996 2.5878 17.2559 2.74408C17.4122 2.90036 17.5 3.11232 17.5 3.33333V16.6667C17.5 16.8877 17.4122 17.0996 17.2559 17.2559C17.0996 17.4122 16.8877 17.5 16.6667 17.5H3.33333C3.11232 17.5 2.90036 17.4122 2.74408 17.2559C2.5878 17.0996 2.5 16.8877 2.5 16.6667V3.33333C2.5 3.11232 2.5878 2.90036 2.74408 2.74408C2.90036 2.5878 3.11232 2.5 3.33333 2.5ZM4.16667 4.16667V15.8333H15.8333V4.16667H4.16667Z"
              fill="#BFCADA"
            />
          </svg>
          <input
            class="input"
            type="checkbox"
          />
        </label>
      </div>
    </div>
    <div
      class="option"
      tabindex="0"
    >
      <div
        class="container"
      >
        <label
          class="container right"
          role="presentation"
        >
          <div
            class="label"
          >
            Three
          </div>
          <svg
            fill="none"
            height="20"
            viewBox="0 0 20 20"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3.33333 2.5H16.6667C16.8877 2.5 17.0996 2.5878 17.2559 2.74408C17.4122 2.90036 17.5 3.11232 17.5 3.33333V16.6667C17.5 16.8877 17.4122 17.0996 17.2559 17.2559C17.0996 17.4122 16.8877 17.5 16.6667 17.5H3.33333C3.11232 17.5 2.90036 17.4122 2.74408 17.2559C2.5878 17.0996 2.5 16.8877 2.5 16.6667V3.33333C2.5 3.11232 2.5878 2.90036 2.74408 2.74408C2.90036 2.5878 3.11232 2.5 3.33333 2.5ZM9.16917 13.3333L15.0608 7.44083L13.8825 6.2625L9.16917 10.9767L6.81167 8.61917L5.63333 9.7975L9.16917 13.3333Z"
              fill="#3385FF"
            />
          </svg>
          <input
            checked=""
            class="input"
            type="checkbox"
          />
        </label>
      </div>
    </div>
  </div>
</div>
`;

exports[`Select supports option deselect 1`] = `
<div
  class="container"
>
  <div
    class="selection"
    tabindex="0"
  >
    <label
      class="container inline"
      role="presentation"
    >
      <div
        class="label"
      >
        One, Three
      </div>
    </label>
    <svg
      class="arrow"
      fill="none"
      height="4"
      viewBox="0 0 8 4"
      width="8"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M4 4L0 0L8 0L4 4Z"
        fill="#a5b1c4"
      />
    </svg>
  </div>
  <div
    class="wrapper"
  >
    <div
      class="option"
      tabindex="0"
    >
      <div
        class="container"
      >
        <label
          class="container right"
          role="presentation"
        >
          <div
            class="label"
          >
            One
          </div>
          <svg
            fill="none"
            height="20"
            viewBox="0 0 20 20"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3.33333 2.5H16.6667C16.8877 2.5 17.0996 2.5878 17.2559 2.74408C17.4122 2.90036 17.5 3.11232 17.5 3.33333V16.6667C17.5 16.8877 17.4122 17.0996 17.2559 17.2559C17.0996 17.4122 16.8877 17.5 16.6667 17.5H3.33333C3.11232 17.5 2.90036 17.4122 2.74408 17.2559C2.5878 17.0996 2.5 16.8877 2.5 16.6667V3.33333C2.5 3.11232 2.5878 2.90036 2.74408 2.74408C2.90036 2.5878 3.11232 2.5 3.33333 2.5ZM9.16917 13.3333L15.0608 7.44083L13.8825 6.2625L9.16917 10.9767L6.81167 8.61917L5.63333 9.7975L9.16917 13.3333Z"
              fill="#3385FF"
            />
          </svg>
          <input
            checked=""
            class="input"
            type="checkbox"
          />
        </label>
      </div>
    </div>
    <div
      class="option"
      tabindex="0"
    >
      <div
        class="container"
      >
        <label
          class="container right"
          role="presentation"
        >
          <div
            class="label"
          >
            Two
          </div>
          <svg
            fill="none"
            height="20"
            viewBox="0 0 20 20"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3.33333 2.5H16.6667C16.8877 2.5 17.0996 2.5878 17.2559 2.74408C17.4122 2.90036 17.5 3.11232 17.5 3.33333V16.6667C17.5 16.8877 17.4122 17.0996 17.2559 17.2559C17.0996 17.4122 16.8877 17.5 16.6667 17.5H3.33333C3.11232 17.5 2.90036 17.4122 2.74408 17.2559C2.5878 17.0996 2.5 16.8877 2.5 16.6667V3.33333C2.5 3.11232 2.5878 2.90036 2.74408 2.74408C2.90036 2.5878 3.11232 2.5 3.33333 2.5ZM4.16667 4.16667V15.8333H15.8333V4.16667H4.16667Z"
              fill="#BFCADA"
            />
          </svg>
          <input
            class="input"
            type="checkbox"
          />
        </label>
      </div>
    </div>
    <div
      class="option"
      tabindex="0"
    >
      <div
        class="container"
      >
        <label
          class="container right"
          role="presentation"
        >
          <div
            class="label"
          >
            Three
          </div>
          <svg
            fill="none"
            height="20"
            viewBox="0 0 20 20"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3.33333 2.5H16.6667C16.8877 2.5 17.0996 2.5878 17.2559 2.74408C17.4122 2.90036 17.5 3.11232 17.5 3.33333V16.6667C17.5 16.8877 17.4122 17.0996 17.2559 17.2559C17.0996 17.4122 16.8877 17.5 16.6667 17.5H3.33333C3.11232 17.5 2.90036 17.4122 2.74408 17.2559C2.5878 17.0996 2.5 16.8877 2.5 16.6667V3.33333C2.5 3.11232 2.5878 2.90036 2.74408 2.74408C2.90036 2.5878 3.11232 2.5 3.33333 2.5ZM9.16917 13.3333L15.0608 7.44083L13.8825 6.2625L9.16917 10.9767L6.81167 8.61917L5.63333 9.7975L9.16917 13.3333Z"
              fill="#3385FF"
            />
          </svg>
          <input
            checked=""
            class="input"
            type="checkbox"
          />
        </label>
      </div>
    </div>
  </div>
</div>
`;

exports[`Select supports single select 1`] = `
<div
  class="container"
>
  <div
    class="selection"
    tabindex="0"
  >
    <label
      class="container inline"
      role="presentation"
    >
      <div
        class="label"
      >
        One
      </div>
    </label>
    <svg
      class="arrow"
      fill="none"
      height="4"
      viewBox="0 0 8 4"
      width="8"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M4 4L0 0L8 0L4 4Z"
        fill="#a5b1c4"
      />
    </svg>
  </div>
  <div
    class="wrapper"
  >
    <div
      class="option"
      tabindex="0"
    >
      <label
        class="container inline"
        role="presentation"
      >
        <div
          class="label"
        >
          One
        </div>
      </label>
    </div>
    <div
      class="option"
      tabindex="0"
    >
      <label
        class="container inline"
        role="presentation"
      >
        <div
          class="label"
        >
          Two
        </div>
      </label>
    </div>
    <div
      class="option"
      tabindex="0"
    >
      <label
        class="container inline"
        role="presentation"
      >
        <div
          class="label"
        >
          Three
        </div>
      </label>
    </div>
  </div>
</div>
`;
